### ISSUE TITLE
[feature request] GPU package power sensor

### ISSUE THREAD

**Issue by NikiSchlifke** [Rem0o/FanControl.Releases#1061]
There is a Powers -> GPU Package reading in Libre Hardware Monitor.

Using this sensor would finally enable consistent and non-oscillating control of fan speed.
The reason obviously would be that cooling requirements are directly proportional to power consumption. 
It's almost the only indicator that's really needed for perfect control with the exception of ambient temperature.

I'd suspect that even  a simple 1:1 Current Power/TDP into fan speed scheme would be more stable than the most complicated setup you can come up with in FanControl.

**Comment by Rem0o** [comment]
This discussion has pop up many times before. 

> would be more stable than the most complicated setup you can come up with in FanControl.

What complicated setup? Just use GPU Core temp, or memory temp, or both, whatever you like.

> The reason obviously would be that cooling requirements are directly proportional to power consumption.

I would argue otherwise. The cooling is required because it just so happens that GPU are inefficient and most of the power consumption is transformed into heat. That heat is then accumulated in the component because the ambient temperature is not cool enough to dissipate that heat fast enough. How much cooling is required is thus a function of the inefficiency of the GPU, power consumption, time and ambient temperature. 

There is a very cool metric that represents all of these at once : the temperature of the GPU.

A simple principle in a linear control system with a feedback loop is that it needs to be bound to the thing it influence/changes. A fan spinning doesn't make power consumption go down. It makes temperature go down.  A quick power spike might not increase your temperature that much. Then there is also the ambient temperature which you need to account for, which power consumption doesn't reflect.

GPU temperature encompass all these things. It's (power * inefficiency) over time radiated/accumulated as heat + ambient temperature.

**Comment by NikiSchlifke** [comment]
Fair enough, it's probably more complicated than I make it out to be here.
However I do think that it would be a very good indicator to have to come up with a more stable control scheme because it's less dependent on other factors. 

The problem with using temperature of the sole indicator is it's prone to positive feedback / instability. A higher temperature requires more fan speed, a higher fan speed makes temperatures go down, requiring less fan speed and so on.
It's a lagging indicator while power consumption is not. 

Sure fans running at speed would increase the power consumption by a small margin, and power consumed by moving electrons might not be turned into heat immediately. However it'd argue the thermodynamic efficiency of a GPU is *essentially* zero.
Meaning almost all the power consumption will end up as heat eventually.

**Comment by Rem0o** [comment]
> The problem with using temperature of the sole indicator is it's prone to positive feedback / instability. A higher temperature requires more fan speed, a higher fan speed makes temperatures go down, requiring less fan speed and so on.
It's a lagging indicator while power consumption is not.

You think temperature is prone to instability issue? Power consumption would be even worse. Take a graph of power over time and temperature, the smoothest of the two will be temperature, trust me.  It doesn't fix this issue.  

In any case, you must set your curve right, with proper response time and hysteresis to get smooth operation.

> Are you open to discussing a bounty on adding it?

Not really, as the software is made to support temperature values. Supporting a new type of unit would need quite a lot of work, especially on the UI side.

Instead of coming out with a solution for me to implement to fix whatever is your problem, why don't you detail your actual config and tell me the problem you have with it. There might be a setting somewhere we can tweak to fix your issue.

**Comment by NikiSchlifke** [comment]
> You think temperature is prone to instability issue? Power consumption would be even worse. Take a graph of power over time and temperature, the smoothest of the two will be temperature, trust me. It doesn't fix this issue.

I really think it would be. The spikey reading of the power consumption should be an issue for sure, but since we already have a time average filter I don't think it's a problem.

> In any case, you must set your curve right, with proper response time and hysteresis to get smooth operation.

I do, FanControl is definitely better than afterburner in that regard, thanks.

> Not really, as the software is made to support temperature values. Supporting a new type of unit would need quite a lot of work, especially on the UI side

That's too bad, can you point me to documentation on how to to implement a custom "File" sensor?

**Comment by Rem0o** [comment]
> That's too bad, can you point me to documentation on how I can implement a custom "File" sensor?
A file sensor simply takes whatever value is inside the file as text, and reports it as a temperature.

In any case, would you care to elaborate your actual problem with your config that made you look for a solution like this? Could you post a screenshot of it?

**Comment by NikiSchlifke** [comment]
> A file sensor simply takes whatever value is inside the file as text, and reports it as a temperature. You can use anything that write to a file to implement one.
That's simple a practical, thanks.


> In any case, would you care to elaborate your actual problem with your config that made you look for a solution like this? Could you post a screenshot of your config?

I'm still experimenting with the program, so far I have been able to get somewhat stable results by using the time average sensor  in conjunction with the auto curve. I also experimented with averaging with the SSD temperature because of the rationale the it might be a decent enough proxy for the case temperature.
If you want to I can email you a screenshot or my userConfig.json

But to be honest I'd rather continue with the debate if you haven't completely made up your mind already. Otherwise we can close this issue.

**Comment by Rem0o** [comment]
Will be open to see your results.

**Comment by NikiSchlifke** [comment]
Okay I have tested it now and it looks promising. Even using power consumption as the sole parameter is feasible, but not optimal.
The main advantage I see is that there is that reaction time to load is potentially shorter. So when you start using the gpu from a cold state the fan speeds up in a constant fashion, depending on the amount of smoothing on the measurements. I chose 6 seconds, so if I go from zero to full load the fan will be at 100% in that time.

That's the setup that works for me. The chained time average is necessary because under spikey load (like neural networks) the power consumption can fluctuate between 100% and 30% over a few seconds. It's less necessary for gaming or rendering.

It's currently not in use because I replaced it with ```scipy.signal.decimate``` which is more aggressive. But if this gets implemented into FanControl I'd suggest patching the Time Average function with a "iterations" field which would function like if you chain several of them together.



To improve this further I'd would want to add temperate back in, at best it would be an offset on the fan speed that is depending on an estimate of the background temperature. But I think something like

speed = power*constant1 + temperatue*constant2 + offset

would work. I guess this can be done with the mix curve in FanControl in average mode and different responses for power and temperature.


This is what the setup looks now. 75% of TDP for 100% fan speed seems about appropriate to me considering only some neural network load can get to 100% for me, rendering and gaming is usually around 60% for me.
<img width="541" alt="fancontrol" src="https://user-images.githubusercontent.com/15075985/188453764-4e7c6937-1e1c-4d68-8e0e-0c0c7d959156.PNG">

The script is here, it smoothes out the value and writes it to a file every second.
``` python
from pathlib import Path
from time import time, sleep
from collections import deque
from scipy import signal
from pynvraw import get_gpus


class Filter(object):
    def __init__(self, first_value, length=60):
        self.data = deque([first_value]*length, maxlen=length)

    def append(self, value):
        self.data.append(value)

    @property
    def value(self):
        filtered = signal.decimate(self.data, q=6, zero_phase=True)
        return float(filtered[-1])


def get_power_consumption():
    gpu = get_gpus()[0]
    return gpu.power


def write_sensor(file_name, value):
    try:
        with Path(file_name).open("w") as file:
            file.write("{}".format(value))
    except EnvironmentError:
        pass

if __name__ == '__main__':
    smooth = Filter(get_power_consumption())
    while True:
        sleep(1 - time() % 1)
        power = get_power_consumption()
        smooth.append(power)
        write_sensor(r"c:\tools\FanControl\power.sensor", smooth.value)
```

**Comment by Rem0o** [comment]
> You think temperature is prone to instability issue? Power consumption would be even worse. 

As expected, you have to average out the power value a few times over a large period to get a "stable enough" value to use, as far as using 5 (!) chained time average cards. 

Let me ask you:  At that point, how is that setup BETTER than simply using the temperature of the GPU as input, which by the way accounts naturally for the ambient temperature in the measurement.

**Comment by NikiSchlifke** [comment]
Really don't want to argue with you like that.

**Comment by Rem0o** [comment]
@NikiSchlifke  Sorry if this seemed harsh/rough. I'm all for experimentation, new ideas, and I welcome users to experiment with complex setups. 

In this case, I had this conversation many times before, and If you got a new take on this, I simply want you to get to the point.

I know power can be used as a measurement for a curve. You went as far as to make it work in a jerry-rigged fashion: great! If you are happy with this, I'm also. 

But streamlining this into the software would require a lot of time to properly integrate this alternative type of input. 
To justify that time, effort and added complexity, I need a clear argument as to why would someone use this over the most simple temperature based setup. 

Basic linear control theory says it's a bad idea and your required setup with a ton of time average cards suggests its a bad idea, while it also cancels one of the only potential benefits, being a non-lagging indicator.

If I'm blinded by bias or I'm clearly missing something, let me know, but this is the situation as I see it from a developer point of view.

**Comment by NikiSchlifke** [comment]
> Basic linear control theory says it's a bad idea and your required setup with a ton of time average cards suggests its a bad idea, while it also cancels one of the only potential benefits, being a non-lagging indicator.

I used very conservative values here to eliminate all fluctuations under heavy fluctuations in neural network load to come up with these values. (It basically drives the gpu into power throttling for 3 seconds then takes 1 second pause and repeat). Much less smoothing is needed for other scenarios. And even in that scenario the lag caused by smoothing is still less than the reaction time inherent with reacting to changes in temperature. 

> If I'm blinded by bias or I'm clearly missing something, let me know, but this is the situation as I see it from a developer point of view.

You've missed my comment about the inherent stability of using power consumption as a measure and the inherent instability / prone to oscillations of using temperature. Any system that can resonate will resonate under certain conditions. You can dampen the oscillation so it's no longer an issue and the hysteresis can be used to eliminate it. Power consumption doesn't have this issue, but different ones it's simply an alternative/supplementary approach. Not worse or better.

> Sorry if this seemed harsh/rough. I'm all for experimentation, new ideas, and I welcome users to experiment with complex setups.

The thing is I didn't feel you were engaging in a proper debate with me because you just repeated your argument of "gpu temperatures includes all those values".


Anyway, it would have been interesting, you mentioned before that the work required on the gui would be more than the potential benefit. I could have helped you out with that. But now I am no longer interested in this project.
I'm gonna turn my python script into a solution that works for me without external dependencies.

I can send you the repo link if you want once I publish it.

**Comment by Rem0o** [comment]
> I can send you the repo link if you want once I publish it.
I will probably link it in the readme. Will give an avenue if some other people want to explore that route.

**Comment by cl0rm** [comment]
I think this would be a great way for case fans.
They ultimately need to remove a certain amount of heat energy from an (almost) closed system. Their heat flow needs to equal (or exceed) the heat flow from the computer parts.
If my GPU gets hot (60C+) its rather likely that at this moment it's dumping 300-400W of thermal energy into the system.
If however my CPU gets hot (even 90-95C), it likely only dumps around 100-200W into the case.
This imbalance of power is rather new in the last ~5 years, before that, GPUs were far less power hungry.
The case fans dont need to keep a temperature stable. The control loop for that is rather complicated/long and unstable. (And which temperature should they even target if we go this route... Average of 75C? Not really, CPU at 25C, GPU at 125C would be that - but rather lethal for the hardware ... ;) ).

Tell me how I could do that without power as a sensor.
10 years ago it was enough to just set the fan curve of the lower case fans to the GPU, upper to the CPU, and call it a day.
With very power-hungry GPUs like my RX6900XT that does not work well. If the GPU dumps the full power, all case fans should (rather quickly) switch to (almost) 100% duty cycle. Otherwise not a chance to dump all the heat.
When it sits in idle, it easily can get to 55-70C, but it will stabilize there, and temperature controlled fan curves will almost always oscillate (GPU is cold, shuts fan off. GPU heats up --> Case fans ramp up considerably --> GPU cools down quickly again as its high temp, but not a continous heat flow of 350W.) This behaivour is anoying when in idle, but needed for a fast enough response while gaming.

Native Power sensors could easily fix this (even with a rather simple 3 point control): Create a combined CPU+GPU Power sensor.
It of course has to be filtered (around 10sec averaging to avoid CPU spikes from triggering it)
If Power < 50W -> Fans 30% (idle, almost silent)
If Power < 150W -> Fans 50% (for mostly CPU only jobs)
Else full Power. 
That way shortly after starting a GPU load, fans increase to make sure the 400W+ total load can adequately be transferred away.

TL;DR: IMHO power-based control is the best way to control CASE fans. (not CPU/GPU). It should be implemented as the underlying drivers (LibreHardwareMonitor) already provide that metric. (For that record, even Load or frequency could be added as a metric. Not overly useful, but it should not be hard)
If the LibreHardwaremonitor sensor plug in would be open source, i would've already done it.
